source/MRMesh/MROpenVDBHelper.h

```diff
@@ -199,6 +199,187 @@ void translateToZero( GredT& grid)
     grid.setTree( outTreePtr );
 }
 
+template <class TreeT, typename Proc>
+class RangeProcessorOne
+{
+public:
+    using InterruptFunc = std::function<bool( void )>;
+    using ProgressFunc = std::function<bool( size_t, size_t )>;
+
+    using ValueT = typename TreeT::ValueType;
+
+    using LeafIterT = typename TreeT::LeafCIter;
+    using TileIterT = typename TreeT::ValueAllCIter;
+    using LeafRange = typename openvdb::tree::IteratorRange<LeafIterT>;
+    using TileRange = typename openvdb::tree::IteratorRange<TileIterT>;
+
+    using TreeAccessor = typename openvdb::tree::ValueAccessor<const TreeT>;
+
+    RangeProcessorOne( const openvdb::math::CoordBBox& b, const TreeT& inT, const Proc& proc ) :
+        mProc( proc ), mBBox( b ), mInTree( inT ), mInAcc( mInTree )
+    {}
+
+    /// Splitting constructor: don't copy the original processor's output tree
+    RangeProcessorOne( RangeProcessorOne& other, tbb::split ) :
+        mProc( other.mProc ),
+        mBBox( other.mBBox ),
+        mInTree( other.mInTree ),
+        mInAcc( mInTree ),
+        mInterrupt( other.mInterrupt ),
+        mProgress( other.mProgress )
+    {}
+
+    void setInterrupt( const InterruptFunc& f ) { mInterrupt = f; }
+    void setProgressFn( const ProgressFunc& f ) { mProgress = f; }
+
+    /// Transform each leaf node in the given range.
+    void operator()( const LeafRange& rCRef )
+    {
+        LeafRange r = rCRef;
+        leafCount = 0;
+        size_t leafCountLast = 0;
+        for ( ; r; ++r )
+        {
+            if ( interrupt() ) break;
+            if ( !( leafCount & 0x400 ) )
+                if ( setProgress( leafCount - leafCountLast, tileCount ) )
```

Comment:
are we sure that it is main thread? (can we setProgress from other threads?)

Reply:
it's not main thread, but real setProgressCallback will definitely only increase (and calls after action with atomic)